<launch>

  <!-- simulation args -->
  <arg name="gazebo_gui"  default="false" />
  <arg name="rviz"        default="true" />
  <arg name="network_viz" default="true"/>

  <!-- patrol demo args -->
  <arg name="task_agent_velocity"    default="1.0"/>  <!-- m/s -->
  <arg name="network_agent_velocity" default="2.0"/>  <!-- m/s -->
  <arg name="sample_count"           default="5000"/>
  <arg name="sample_variance"        default="2.0"/>
  <arg name="margin1"                default="0.05"/> <!-- [0 1] -->
  <arg name="confidence1"            default="0.70"/> <!-- (0.5 1) -->
  <arg name="margin2"                default="0.05"/> <!-- [0 1] -->
  <arg name="confidence2"            default="0.70"/> <!-- (0.5 1) -->
  <arg name="collision_distance"     default="4.0"/>  <!-- m -->

  <!-- patrol demo params -->
  <param name="/task_agent_count"    value="2"/> <!-- must match number of agents spawned! -->
  <param name="/network_agent_count" value="4"/> <!-- must match number of agents spawned! -->
  <param name="/patrol_radius"       value="40.0"/>
  <param name="/desired_altitude"    value="2.0"/>

  <!-- heatmap demo params -->
  <param name="/heatmap_x_size"       value="40.0"/>
  <param name="/heatmap_y_size"       value="40.0"/>
  <param name="/heatmap_lines"       value="20.0"/>

  <!-- channel rate model params -->
  <param name="/N0" value="-70.0"/> <!-- -85.0  receiver noise (dBm) -->
  <param name="/n"  value="2.52"/>  <!--   2.52 decay exponent -->
  <param name="/L0" value="-48.0"/> <!-- -53.0  transmit power (dBm) -->
  <param name="/a"  value="0.2"/>   <!--   0.2  variance sigmoid param 1 -->
  <param name="/b"  value="0.6"/>   <!--   0.6  variance sigmoid param 2 -->

  <!-- gazebo and rviz -->
  <include file="$(find intel_aero_gazebo)/launch/simulation.launch">
    <arg name="gazebo_gui"       value="$(arg gazebo_gui)"/>
    <arg name="world"            value="worlds/empty.world"/>
    <arg name="rviz"             value="$(arg rviz)"/>
    <arg name="rviz_config_file" value="$(find intel_aero_demos)/rviz/patrol.rviz"/>
  </include>

  <!-- load saved map -->
  <node pkg="map_server"
        type="map_server"
        name="static_map_node"
        args="$(find intel_aero_gazebo)/maps/empty.yaml">
    <remap from="map"      to="/projected_map"/>
    <param name="frame_id" value="world"/>
  </node>

  <!-- network topology visualization -->
  <node pkg="network_status"
        type="routing_channel_visualization_node"
        name="routing_channel_visualization_node"
        output="screen">
    <param name="world_frame" value="world"/>
  </node>

  <!-- network planner node -->
  <include file="$(find network_planner)/launch/patrol.launch">
    <arg name="max_velocity"       value="$(arg network_agent_velocity)"/>
    <arg name="sample_count"       value="$(arg sample_count)"/>
    <arg name="sample_variance"    value="$(arg sample_variance)"/>
    <arg name="costmap_topic"      value="/projected_map"/>
    <arg name="margin1"            value="$(arg margin1)"/>
    <arg name="confidence1"        value="$(arg confidence1)"/>
    <arg name="margin2"            value="$(arg margin2)"/>
    <arg name="confidence2"        value="$(arg confidence2)"/>
    <arg name="collision_distance" value="$(arg collision_distance)"/>
  </include>

  <!-- load saved octomap
  <node pkg="octomap_server"
        type="octomap_server_node"
        name="static_map_node"
        args="$(find intel_aero_gazebo)/maps/spawc_no_ground.bt"
        output="screen">
    <remap from="projected_map" to="/projected_map"/>
    <param name="frame_id" value="world"/>
  </node>
 -->

  <!-- task agents -->

  <group>
    <arg name="id" value="1"/>
    <include file="$(find intel_aero_gazebo)/launch/intel_aero_spawn_only.launch">
      <arg name="id"             value="$(arg id)"/>
      <arg name="x"              value="20.0"/>
      <arg name="y"              value="20.0"/>
      <arg name="linear_vel_des" value="$(arg task_agent_velocity)"/>
    </include>
  </group>

  <group>
    <arg name="id" value="2"/>
    <include file="$(find intel_aero_gazebo)/launch/intel_aero_spawn_only.launch">
      <arg name="id"             value="$(arg id)"/>
      <arg name="x"              value="0.0"/>
      <arg name="y"              value="0.0"/>
      <arg name="linear_vel_des" value="$(arg task_agent_velocity)"/>
    </include>
  </group>

  <!-- network agents -->

  <group>
    <arg name="id" value="3"/>
    <include file="$(find intel_aero_gazebo)/launch/intel_aero_spawn_only.launch">
      <arg name="id"             value="$(arg id)"/>
      <arg name="x"              value="10.0"/>
      <arg name="y"              value="10.0"/>
    </include>
  </group>

  <group>
    <arg name="id" value="4"/>
    <include file="$(find intel_aero_gazebo)/launch/intel_aero_spawn_only.launch">
      <arg name="id"             value="$(arg id)"/>
      <arg name="x"              value="-10.0"/>
      <arg name="y"              value="10.0"/>
    </include>
  </group>

  
  <group>
    <arg name="id" value="5"/>
    <include file="$(find intel_aero_gazebo)/launch/intel_aero_spawn_only.launch">
      <arg name="id"             value="$(arg id)"/>
      <arg name="x"              value="10.0"/>
      <arg name="y"              value="-10.0"/>
    </include>
  </group>

  <group>
    <arg name="id" value="6"/>
    <include file="$(find intel_aero_gazebo)/launch/intel_aero_spawn_only.launch">
      <arg name="id"             value="$(arg id)"/>
      <arg name="x"              value="-10.0"/>
      <arg name="y"              value="-10.0"/>
    </include>
  </group>
  

</launch>
